File: 5-最长回文数组（dp）.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        if(n < 2)
            return s;
        else{
            vector<vector<bool>> dp(n,vector<bool>(n,false));//最小长度为1，防止最后返回时不返回子串
            int maxlen = 1, begin = 0;
            int i = 0, j = 0, l = 0;
            for(i = 0; i< n; i++)
                dp[i][i] = true;
            for(l = 2; l<=n; l++){
                for(i = 0;i<n;i++){
                    int r = i + l - 1;
                    if(r>=n)
                        break;
                    if(s[i] == s[r]){
                        if(r - i < 3)//长度不足3时才置为真，否则与去头尾的子串一致
                            dp[i][r] = true;
                        else
                            dp[i][r] = dp[i + 1][r - 1];
                    }
                    else
                        dp[i][r] = false;
                    if(dp[i][r]&&r-i + 1>maxlen){
                        begin = i;
                        maxlen = r - i + 1;
                    }
                }
            }
            return s.substr(begin,maxlen);
        }
    }
};

File: 5-最长回文数组（dp）.java
class Solution {
    public String longestPalindrome(String s) {
        int n = s.length();
        int begin = 0;
        int len = 1;
        if(n < 2)
            return s;
        int l1 = 0, r1 = 0;
        boolean [][] dp = new boolean[n][n];
        for(int i = 0; i < n; i++){
            dp[i][i] = true;
        }
        for(int i = 2; i <= n; i++){
            for(int j = 0; j < n; j++){
                int r = i + j - 1;
                if(r>=n)
                    break;
                if(s.charAt(j) == s.charAt(r)){
                    if(r - j < 3){
                        dp[j][r] = true;
                    }
                    else{
                        dp[j][r] = dp[j+1][r-1];
                    }
                }
                else
                    dp[j][r] = false;
                if(dp[j][r] == true && r-j+1 > len){
                    len = r-j+1;
                    begin = j;
                }
            }
        }
        return s.substring(begin, begin + len);

    }
}